### ISSUE TITLE
Product indexer + queue extremely slow

### ISSUE THREAD

**Issue by peterjaap** [algolia/algoliasearch-magento-2#426]
When the queue/cron is disabled, the product indexer gets fed the `$productIds` array, which contains **all** product ID's. We have a shop with 100.000+ SKU's of which only about 150 products are actually visible on the frontend.

When I look in Algolia, I only see those ~150 products. So my guess is that the rest is ignored. Running a full reindex (for good reasons ;)) takes about 6 hours. Filtering the `$productIds` array on only products that have visibility set to > 1 should then be taking into the indexer, considerably cutting down the reindexing time.

Am I missing something here in my assumptions? Maybe something to do with the configurable variations?

**Comment by peterjaap** [comment]
Ok that didn't really have any effect. When I log the `$productIds` variable in the [execute](https://github.com/algolia/algoliasearch-magento-2/blob/master/Model/Indexer/Product.php#L46) method, I see it printed out 4 times when I run it through `php bin/magento ind:reindex algolia_runner`. Why is that?

**Comment by peterjaap** [comment]
Ok so the problem isn't with the visibility. The indexer is just generally slow (at least here). If I want to force reindex all my ~150 products, it takes about 6 hours.

But I just found out that when I push all ids and let the indexer chunk it into chunks of 30 items each (or whatever), it's slow. But when I 'chunk' it into chunks of 1 item, each item takes about 10 seconds. So 150 products * 4 storeviews * 10 seconds = 6000 seconds which is about 1.5 hours. Better than 6 hours, but still too slow for what is basically just a simple JSON object....

**Comment by peterjaap** [comment]
Actually, it is also very slow when I insert them with chunk 1 into the database and then let the runner run. I now insert them 1 by 1 into the database and run the queue directly after inserting one so the queue runner only has to process one item. I feel like there is a memory leak or something somewhere.

**Comment by peterjaap** [comment]
In case anyone is interested in my fugly-hack-until-it-works-and-dont-refactor script;

```
<?php

$storeviewIds = [1, 2];
$productIds = [1,2,3,4];

foreach ($storeviewIds as $storeviewId) {
    foreach ($productIds as $productId) {
        echo "echo \"INSERT INTO algoliasearch_queue SET created=NOW(), class = 'Algolia\\\\\\AlgoliaSearch\\\\\\Helper\\\\\\Data', method = 'rebuildStoreProductIndex', data = '{\\\"store_id\\\":".$storeviewId.",\\\"product_ids\\\":[".$productId."]}', max_retries = 3, data_size =1;\" | mysql $DBNAME && EMPTY_QUEUE=1  php -dmemory_limit=-1 bin/magento ind:reindex algolia_queue_runner" . PHP_EOL;
    }
}
```

```
DBNAME=yourdbname php generate.php > commands.sh && sh commands.sh
```

**Comment by JanPetr** [comment]
Hello Peter,

thank you for creating the issue!

To answer your first question why Algolia is processing all 100k products instead of only 150 with disabled queue: With disabled queue, Algolia is fetching all products in order to delete those not visible or disabled products. If the extension doesn't fetch them, it might happen those products will stay in Algolia indices and remain searchable.
With the queue enabled, there is a different reindexing mechanism which reindexes only visible products to a separate index and when the reindexing is done, the new index replaces the old one trashing all outdated products. Is that understandable?

About 10 seconds time to reindex a single product - I'm not able to produce. On my Magento instances it takes not even a second. (But by this, I'm not saying it's perfect.)
Did you manage to investigate what takes the most time during creating the Algolia record?

**Comment by peterjaap** [comment]
We're planning to Blackfire this issue soon. Will report back.

**Comment by peterjaap** [comment]
I added some logging in `ProductHelper.php` and it seems retrieving the addtional attribute data slows down my indexing process;

```
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN CREATE RECORD 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN TRANSPORT 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO: <<<<< END TRANSPORT 475588 1 (Client.nl) (5.793571472168E-5sec) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN VISIBILITY 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO: <<<<< END VISIBILITY 475588 1 (Client.nl) (3.0040740966797E-5sec) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN CUSTOM DATA ARRAY 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO: <<<<< END CUSTOM DATA ARRAY 475588 1 (Client.nl) (0.026535034179688sec) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN ADD ATTRIBUTES 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN ATTRIBUTE description 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO: <<<<< END ATTRIBUTE description 475588 1 (Client.nl) (3.9100646972656E-5sec) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN ATTRIBUTE ordered_qty 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO: <<<<< END ATTRIBUTE ordered_qty 475588 1 (Client.nl) (2.598762512207E-5sec) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN ATTRIBUTE total_ordered 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO: <<<<< END ATTRIBUTE total_ordered 475588 1 (Client.nl) (2.3126602172852E-5sec) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN ATTRIBUTE rating_summary 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO: <<<<< END ATTRIBUTE rating_summary 475588 1 (Client.nl) (2.9087066650391E-5sec) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN ADD CATEGORY DATA 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO: <<<<< END ADD CATEGORY DATA 475588 1 (Client.nl) (0.034063100814819sec) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN ADD IMAGE DATA 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO: <<<<< END ADD IMAGE DATA 475588 1 (Client.nl) (0.032708168029785sec) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN ADD IN STOCK 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO: <<<<< END ADD IN STOCK 475588 1 (Client.nl) (0.0059690475463867sec) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN ADD STOCK QTY 475588 1 (Client.nl) [] []
[2018-03-27 15:25:08] main.INFO: <<<<< END ADD STOCK QTY 475588 1 (Client.nl) (4.1007995605469E-5sec) [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO:  [] []
[2018-03-27 15:25:08] main.INFO: >>>>> BEGIN GET SUBPRODUCTS 475588 1 (Client.nl) [] []
[2018-03-27 15:25:10] main.INFO: <<<<< END GET SUBPRODUCTS 475588 1 (Client.nl) (1.9285910129547sec) [] []
[2018-03-27 15:25:10] main.INFO:  [] []
[2018-03-27 15:25:10] main.INFO:  [] []
[2018-03-27 15:25:10] main.INFO: >>>>> BEGIN SET ADDITIONAL ATTRIBUTES 475588 1 (Client.nl) [] []
[2018-03-27 15:25:29] main.INFO: <<<<< END SET ADDITIONAL ATTRIBUTES 475588 1 (Client.nl) (18.858809947968sec) [] []
[2018-03-27 15:25:29] main.INFO:  [] []
[2018-03-27 15:25:29] main.INFO:  [] []
[2018-03-27 15:25:29] main.INFO: >>>>> BEGIN ADD PRICE DATA 475588 1 (Client.nl) [] []
[2018-03-27 15:25:34] main.INFO: <<<<< END ADD PRICE DATA 475588 1 (Client.nl) (5.2823140621185sec) [] []
[2018-03-27 15:25:34] main.INFO: <<<<< END ADD ATTRIBUTES 475588 1 (Client.nl) (26.226011037827sec) [] []
[2018-03-27 15:25:34] main.INFO:  [] []
[2018-03-27 15:25:34] main.INFO:  [] []
[2018-03-27 15:25:34] main.INFO: >>>>> BEGIN EVENT algolia_subproducts_index 475588 1 (Client.nl) [] []
[2018-03-27 15:25:34] main.INFO: <<<<< END EVENT algolia_subproducts_index 475588 1 (Client.nl) (0.00020408630371094sec) [] []
[2018-03-27 15:25:34] main.INFO:  [] []
[2018-03-27 15:25:34] main.INFO:  [] []
[2018-03-27 15:25:34] main.INFO: >>>>> BEGIN EVENT algolia_after_create_product_object 475588 1 (Client.nl) [] []
[2018-03-27 15:25:34] main.INFO: <<<<< END EVENT algolia_after_create_product_object 475588 1 (Client.nl) (6.6995620727539E-5sec) [] []
[2018-03-27 15:25:34] main.INFO: <<<<< END CREATE RECORD 475588 1 (Client.nl) (26.299291849136sec) [] []
```

**Comment by peterjaap** [comment]
Dove into 'additional attributes' and found that especially size and color attributes take a long time;

```
[2018-03-27 15:36:05] main.INFO: >>>>> BEGIN SET ADDITIONAL ATTRIBUTES 558063 1 (Client.nl) [] []
[2018-03-27 15:36:05] main.INFO:  [] []
[2018-03-27 15:36:05] main.INFO:  [] []
[2018-03-27 15:36:05] main.INFO: >>>>> BEGIN GETTING RESOURCE [] []
[2018-03-27 15:36:05] main.INFO:  [] []
[2018-03-27 15:36:05] main.INFO:  [] []
[2018-03-27 15:36:05] main.INFO: >>>>> BEGIN GETTING RESOURCE [] []
[2018-03-27 15:36:05] main.INFO: <<<<< END GETTING RESOURCE (9.0122222900391E-5sec) [] []
[2018-03-27 15:36:05] main.INFO:  [] []
[2018-03-27 15:36:05] main.INFO:  [] []
[2018-03-27 15:36:05] main.INFO: >>>>> BEGIN GETTING ATTRIBUTE VALUE color [] []
[2018-03-27 15:36:05] main.INFO: <<<<< END GETTING ATTRIBUTE VALUE color (1.5020370483398E-5sec) [] []
[2018-03-27 15:36:05] main.INFO:  [] []
[2018-03-27 15:36:05] main.INFO:  [] []
[2018-03-27 15:36:05] main.INFO: >>>>> BEGIN ADD NON NULL VALUE color [] []
[2018-03-27 15:36:05] main.INFO: <<<<< END ADD NON NULL VALUE color (4.0531158447266E-6sec) [] []
[2018-03-27 15:36:05] main.INFO:  [] []
[2018-03-27 15:36:05] main.INFO:  [] []
[2018-03-27 15:36:05] main.INFO: >>>>> BEGIN ADD NULL VALUE color [] []
[2018-03-27 15:36:25] main.INFO: <<<<< END ADD NULL VALUE color (19.501872062683sec) [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN GETTING RESOURCE [] []
[2018-03-27 15:36:25] main.INFO: <<<<< END GETTING RESOURCE (0.00010085105895996sec) [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN GETTING ATTRIBUTE VALUE sku [] []
[2018-03-27 15:36:25] main.INFO: <<<<< END GETTING ATTRIBUTE VALUE sku (2.3126602172852E-5sec) [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN ADD NON NULL VALUE sku [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN GETTING RESOURCE [] []
[2018-03-27 15:36:25] main.INFO: <<<<< END GETTING RESOURCE (9.2029571533203E-5sec) [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN GETTING ATTRIBUTE VALUE price [] []
[2018-03-27 15:36:25] main.INFO: <<<<< END GETTING ATTRIBUTE VALUE price (1.4066696166992E-5sec) [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN ADD NON NULL VALUE price [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN GETTING RESOURCE [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN GETTING RESOURCE [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN GETTING RESOURCE [] []
[2018-03-27 15:36:25] main.INFO: <<<<< END GETTING RESOURCE (7.8916549682617E-5sec) [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN GETTING ATTRIBUTE VALUE created_at [] []
[2018-03-27 15:36:25] main.INFO: <<<<< END GETTING ATTRIBUTE VALUE created_at (1.5974044799805E-5sec) [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN ADD NON NULL VALUE created_at [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN GETTING RESOURCE [] []
[2018-03-27 15:36:25] main.INFO: <<<<< END GETTING RESOURCE (9.0122222900391E-5sec) [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN GETTING ATTRIBUTE VALUE size [] []
[2018-03-27 15:36:25] main.INFO: <<<<< END GETTING ATTRIBUTE VALUE size (1.6927719116211E-5sec) [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN ADD NON NULL VALUE size [] []
[2018-03-27 15:36:25] main.INFO: <<<<< END ADD NON NULL VALUE size (2.8610229492188E-6sec) [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO:  [] []
[2018-03-27 15:36:25] main.INFO: >>>>> BEGIN ADD NULL VALUE size [] []
[2018-03-27 15:36:31] main.INFO: <<<<< END ADD NULL VALUE size (6.2481491565704sec) [] []
[2018-03-27 15:36:31] main.INFO:  [] []
[2018-03-27 15:36:31] main.INFO:  [] []
[2018-03-27 15:36:31] main.INFO: >>>>> BEGIN GETTING RESOURCE [] []
[2018-03-27 15:36:31] main.INFO: <<<<< END GETTING RESOURCE (9.6797943115234E-5sec) [] []
[2018-03-27 15:36:31] main.INFO:  [] []
[2018-03-27 15:36:31] main.INFO:  [] []
[2018-03-27 15:36:31] main.INFO: >>>>> BEGIN GETTING ATTRIBUTE VALUE sock_type [] []
[2018-03-27 15:36:31] main.INFO: <<<<< END GETTING ATTRIBUTE VALUE sock_type (1.6212463378906E-5sec) [] []
[2018-03-27 15:36:31] main.INFO:  [] []
[2018-03-27 15:36:31] main.INFO:  [] []
[2018-03-27 15:36:31] main.INFO: >>>>> BEGIN ADD NON NULL VALUE sock_type [] []
[2018-03-27 15:36:31] main.INFO: <<<<< END ADD NON NULL VALUE sock_type (4.0531158447266E-6sec) [] []
[2018-03-27 15:36:31] main.INFO:  [] []
[2018-03-27 15:36:31] main.INFO:  [] []
[2018-03-27 15:36:31] main.INFO: >>>>> BEGIN ADD NULL VALUE sock_type [] []
[2018-03-27 15:36:38] main.INFO: <<<<< END ADD NULL VALUE sock_type (6.8003499507904sec) [] []
[2018-03-27 15:36:38] main.INFO:  [] []
[2018-03-27 15:36:38] main.INFO:  [] []
[2018-03-27 15:36:38] main.INFO: >>>>> BEGIN GETTING RESOURCE [] []
[2018-03-27 15:36:38] main.INFO: <<<<< END GETTING RESOURCE (9.608268737793E-5sec) [] []
[2018-03-27 15:36:38] main.INFO:  [] []
[2018-03-27 15:36:38] main.INFO:  [] []
[2018-03-27 15:36:38] main.INFO: >>>>> BEGIN GETTING ATTRIBUTE VALUE brand [] []
[2018-03-27 15:36:38] main.INFO: <<<<< END GETTING ATTRIBUTE VALUE brand (1.5974044799805E-5sec) [] []
[2018-03-27 15:36:38] main.INFO:  [] []
[2018-03-27 15:36:38] main.INFO:  [] []
[2018-03-27 15:36:38] main.INFO: >>>>> BEGIN ADD NON NULL VALUE brand [] []
[2018-03-27 15:36:38] main.INFO: <<<<< END SET ADDITIONAL ATTRIBUTES 558063 1 (Client.nl) (32.705775976181sec) [] []
```

**Comment by peterjaap** [comment]
So it seems the foreach loop in the `addNullValue` method loops over all subproducts. Since we have up to 10.000 subproducts for one product, this obviously takes a long time. Will try to devise of a more efficient method.

**Comment by peterjaap** [comment]
@JanPetr maybe you can help me explain what `addNullValue` exactly does. It loops through all products and when the product is not in stock, it skips it.

Otherwise, it retrieves the attribute text value for the attribute it is currently looping over and then adds all the values to a unique'd array and passes that back to the main loop.

Meanwhile, also `subProductImages` is filled (not really single responsibility?). Why exactly is all of this necessary?

An optimization would be to fetch all in-stock simple products and create an array on initialization of the indexer, then retrieve the necessary data from that in-memory array.

**Comment by peterjaap** [comment]
So I have discovered the `subProductImages` array is filled with images that are attached to the simple products. Presumably to fill a swatch somewhere? I can't find where in Algolia this shows up.

But I don't understand these lines;

```
        if (is_array($values) && count($values) > 0) {
            $customData[$attribute['attribute']] = array_values(array_unique($values));
        }
```

When I log `$customData`, the array contains info from a configurable product, but the attribute is filled with only one value, like this;

```
    [color] => Array
        (
            [0] => Bronze
        )
    [size] => Array
        (
            [0] => 7
        )
```

Shouldn't this contain at least all values of the simples? And if not, why only one value? I have commented out the `addNullValue()` method while testing and it seems it indexes and works just fine. Am I missing something?

**Comment by peterjaap** [comment]
I think you meant it to contain all attribute values of the subproducts for filtering purposes. I must've ran into a bug? `$values` is defined as an empty var [here](https://github.com/algolia/algoliasearch-magento-2/blob/master/Helper/Entity/ProductHelper.php#L665) but also [here](https://github.com/algolia/algoliasearch-magento-2/blob/master/Helper/Entity/ProductHelper.php#L699).

My guess is the second `$values = [];` (inside `getValues()`) should be removed and `$values` should be handed to `getValues()` as an argument from `addNullValue()`. That way, the code would make sense, although not very efficient.

**Comment by peterjaap** [comment]
I did a quick hack to temporarily fix this for us. Since we don't use images on simple product level, I've ignored `$subProductImages` for now.

I've DI'd a database connection into `ProductHelper` and rewrote the `addNullValue()` method to this (which only handles dropdown attributes for now, since we only use those for now);

```
    private function addNullValue($customData, $subProducts, $storeId, $attribute, AttributeResource $attributeResource)
    {
        $values = [];

        /** @var \Magento\Catalog\Model\ResourceModel\Product\Collection $subProducts */
        $subProductIds = $subProducts->getAllIds();

        /** $values should contain all attribute values from given subProducts */
        $attributeInfo = $this->db->fetchRow(
            $this->db->select()
                ->from('eav_attribute')
                ->where('entity_type_id = ?', 4)
                ->where('attribute_code = ?', $attribute['attribute'])
        );

        if ($attributeInfo['backend_type'] == 'int') {
            $values = $this->db->fetchAll($this->db->select()
                ->distinct(true)
                ->from(['cpe' => $this->resource->getTableName('catalog_product_entity_' . $attributeInfo['backend_type'])], 'value')
                ->joinInner(['eaov' => $this->resource->getTableName('eav_attribute_option_value')], 'cpe.value = eaov.option_id', ['text_value' => 'value'])
                ->where('cpe.attribute_id = ?', $attributeInfo['attribute_id'])
                ->where('eaov.store_id = ?', $storeId)
                ->where('entity_id IN (?)', $subProductIds));

            $values = array_column($values, 'text_value');
        }

        if (is_array($values) && count($values) > 0) {
            $customData[$attribute['attribute']] = array_values(array_unique($values));
        }

        return $customData;
    }
```

This cuts down processing time for us from 6 hours to 15 minutes.

**Comment by JanPetr** [comment]
Hello @peterjaap!

Thank you very much for your hard work and intensive investigation.
What you have found out definitely makes sense and you're right - there is a bug, which fetches always all sub products and the extension has to iterate over everything over and over again.

I create a fix which should prevent that - https://github.com/algolia/algoliasearch-magento-2/pull/462
The extension won't fetch all sub products, but if it's set you don't want to display out of stock options, it'll fetch only sub products which should be displayed.

Could you implement the fix, give it a try and let me know if it works OK on your side?

Thanks a bunch! 💯

**Comment by peterjaap** [comment]
Awesome, that worked! Processing time is now 25 minutes, much better and I can live with this. Thanks :+1:

**Comment by JanPetr** [comment]
Thanks @peterjaap!
The `$values` issue has been fixed in the same PR (#462), so I'm closing the issue now.

Thanks a lot for your feedback 💯 Feel free to grab [some swag](https://docs.google.com/a/algolia.com/forms/d/e/1FAIpQLScFHcqkzw8QoS-XkU1zZ4eOg77K2t88qIXrqOC-828ahyiCnQ/viewform), if you like 🚀
